src/Dialect/ONNX/MLIRDialectBuilder.cpp

```diff
@@ -164,49 +167,76 @@ Value MathBuilder::select(Value cmp, Value lhs, Value rhs) const {
 }
 
 Value MathBuilder::constant(Type type, double val) const {
-  Attribute constantAttr;
+  Value constant = nullptr;
   TypeSwitch<Type>(type)
-      .Case<Float16Type>(
-          [&](Type) { constantAttr = b.getF16FloatAttr((float)val); })
-      .Case<Float32Type>(
-          [&](Type) { constantAttr = b.getF32FloatAttr((float)val); })
-      .Case<Float64Type>(
-          [&](Type) { constantAttr = b.getF64FloatAttr((float)val); })
-      .Case<IntegerType>([&](Type) {
+      .Case<Float16Type>([&](Type) {
+        constant = b.create<arith::ConstantOp>(loc, b.getF16FloatAttr(val));
+      })
+      .Case<Float32Type>([&](Type) {
+        constant = b.create<arith::ConstantOp>(loc, b.getF32FloatAttr(val));
+      })
+      .Case<Float64Type>([&](Type) {
+        constant = b.create<arith::ConstantOp>(loc, b.getF64FloatAttr(val));
+      })
+      .Case<IntegerType>([&](IntegerType type) {
         assert(val == (int64_t)val && "value is ambiguous");
-        auto width = type.cast<IntegerType>().getWidth();
-        if (width == 1) {
-          constantAttr = b.getBoolAttr(val != 0);
-        } else {
-          constantAttr = b.getIntegerAttr(type, APInt(width, (int64_t)val));
+        unsigned width = type.getWidth();
+
+        if (width == 1)
+          constant = b.create<arith::ConstantOp>(loc, b.getBoolAttr(val != 0));
+        else {
+          assert(type.isSignless() &&
+                 "arith::ConstantOp requires a signless type.");
+          constant = b.create<arith::ConstantOp>(
+              loc, b.getIntegerAttr(type, APInt(width, (int64_t)val)));
         }
       })
-      .Case<IndexType>(
-          [&](Type) { constantAttr = b.getIntegerAttr(type, (int64_t)val); })
+      .Case<IndexType>([&](Type) {
+        constant =
+            b.create<arith::ConstantOp>(loc, b.getIntegerAttr(type, val));
+      })
       .Default([](Type) { llvm_unreachable("unsupported element type"); });
-  return b.create<arith::ConstantOp>(loc, constantAttr);
+
+  assert(constant != nullptr && "Expecting valid constant value");
+  return constant;
 }
 
 Value MathBuilder::constantIndex(int64_t val) const {
   Attribute constantAttr = b.getIntegerAttr(b.getIndexType(), val);
   return b.create<arith::ConstantOp>(loc, constantAttr);
 }
 
-// For some reason, operations on unsigned int are often unhappy because
-// operations are mainly used on signless integers. So this cast remove the sign
-// of unsigned int for successful processing, to the best of my understanding.
+Value MathBuilder::createArithICmp(
```

Comment:
Could we overload `createArithCmp` (dropping the `I`) with the `arith::CmpIPredicate` for float, and a separate overloaded call for `arith::CmpFPredicate`

Reply:
done